### PR TITLE
feat(skymp5-server): revert remove original NPCs spawn & add npc settings

### PR DESCRIPTION
**Pull request by che6a**
closes #484 #1163

### CODE REVIEW

File: skymp5-server/cpp/server_guest_lib/WorldState.cpp [hunk 1]
```diff
@@ -252,11 +253,11 @@ bool WorldState::AttachEspmRecord(const espm::CombineBrowser& br,
                                   const espm::IdMapping& mapping)
 {
   auto& cache = GetEspmCache();
-  auto refr = reinterpret_cast<espm::REFR*>(record);
+  auto* refr = espm::Convert<espm::REFR>(record);
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/WorldState.cpp:256]: please keep using reinterpret_cast here and add a comment about why:
This is a hot path, we prefer reinterpret_cast since it doesn't generate any assembler code

File: skymp5-server/cpp/server_guest_lib/WorldState.h [hunk 2]
```diff
@@ -168,6 +169,21 @@ class WorldState
     lastMovUpdateByIdx;
 
   bool isPapyrusHotReloadEnabled = false;
+  bool npcEnabled;
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/WorldState.h:172]: please init values here

File: skymp5-server/cpp/server_guest_lib/WorldState.cpp [hunk 3]
```diff
@@ -59,6 +59,7 @@ struct WorldState::Impl
 };
 
 WorldState::WorldState()
+  : npcEnabled(false)
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/WorldState.cpp:62]: please don't init this value here

File: skymp5-server/cpp/server_guest_lib/WorldState.h [hunk 4]
```diff
@@ -154,6 +154,7 @@ class WorldState
                      std::chrono::system_clock::duration dur);
   std::optional<std::chrono::system_clock::duration> GetRelootTime(
     std::string recordType) const;
+  void AddLoadOrderElement(std::string fileName);
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/WorldState.h:157]: this is not needed. you can use the existing variable espmFiles.

index in this vector == first 2 digits of form id

**Review by Pospelove — CHANGES_REQUESTED** [review]
1. take a look at worldOrCell. if you can find espm::CELL then it is an interior, otherwise it is not
2. it'd be nice to have detailed settings like:

```json
{
"npcSettings": {
"default": { "spawnInInterior": true, "spawnInExterior": false },
"SweetPie.esp": {"spawnInInterior": false, "spawnInExterior": true}
}
}
```
3. please add docs for this new option

File: docs/docs_server_configuration_reference.md [hunk 5]
```diff
@@ -248,3 +248,42 @@ Allows tuning settings related to in-game chat, such as message visibility radiu
...
+{
+  // ...
+  "npcSettings": {
+    "Skyrim.esm": {
+      "spawnInInterior": true,
+      "spawnInExterior": false
+    },
+    "Dawnguard.esm": {
+      "spawnInInterior": false,
+      "spawnInInterior": true
```
Comment (Pospelove) [inline docs/docs_server_configuration_reference.md:280]: typo

File: skymp5-server/cpp/server_guest_lib/WorldState.h [hunk 6]
```diff
@@ -39,6 +39,13 @@ class WorldState
 public:
   using FormCallbacksFactory = std::function<FormCallbacks()>;
 
+  struct NpcSettingsEntry
+  {
+    bool spawnInInterior = true;
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/WorldState.h:44]: let's make false

like it's a default value of struct isnt it?

File: skymp5-server/cpp/server_guest_lib/WorldState.cpp [hunk 7]
```diff
@@ -652,3 +701,32 @@ std::optional<std::chrono::system_clock::duration> WorldState::GetRelootTime(
...
+bool WorldState::NpcSourceFilesOverriden() const noexcept
+{
+  return !npcSettings.empty();
+}
+
+bool WorldState::IsNpcAllowed(uint32_t baseId) const noexcept
+{
+  uint32_t npcFileIdx = GetFileIdx(baseId);
+  for (const auto& [fileName, _] : npcSettings) {
+    auto it = std::find(espmFiles.begin(), espmFiles.end(), fileName);
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/WorldState.cpp:714]: what if `it` would be espmfiles.end()

File: skymp5-server/cpp/server_guest_lib/WorldState.cpp [hunk 8]
```diff
@@ -293,6 +330,18 @@ bool WorldState::AttachEspmRecord(const espm::CombineBrowser& br,
     return false;
   }
 
+  espm::LookupResult res = br.LookupById(worldOrCell);
+  uint16_t cellFlags =
+    reinterpret_cast<espm::CELL*>(res.rec)->GetData(cache).flags;
+  bool isInterior = cellFlags & espm::CELL::Flags::Interior;
+  bool isExterior = !isInterior;
+  const NpcSettingsEntry& entry = npcSettings[espmFiles[GetFileIdx(baseId)]];
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/WorldState.cpp:338]: maybe would be better to not rely on default values here, but use iterators instead. and if not found in npcSettings, set needed values explicitly.

espmFiles may also not contain such index for some weird reason.

let's add some checks and logging

File: skymp5-server/cpp/addon/ScampServer.cpp [hunk 9]
```diff
@@ -146,6 +148,31 @@ ScampServer::ScampServer(const Napi::CallbackInfo& info)
                    spdlog::level::to_string_view(logger->level()));
     }
 
+    if (serverSettings.find("npcEnabled") != serverSettings.end()) {
+      partOne->worldState.npcEnabled =
+        serverSettings.at("npcEnabled").get<bool>();
+    }
+
+    if (serverSettings.find("npcSettings") != serverSettings.end()) {
```
Comment (Pospelove) [inline skymp5-server/cpp/addon/ScampServer.cpp:156]: let's add some logs to simplify life of SupAidmi

**Review by Pospelove — CHANGES_REQUESTED** [review]

File: skymp5-server/cpp/server_guest_lib/WorldState.cpp [hunk 10]
```diff
@@ -345,8 +345,8 @@ bool WorldState::AttachEspmRecord(const espm::CombineBrowser& br,
 
   if (isNpc) {
     espm::LookupResult res = br.LookupById(worldOrCell);
-    uint16_t cellFlags =
-      reinterpret_cast<espm::CELL*>(res.rec)->GetData(cache).flags;
+    auto cellRecord = reinterpret_cast<espm::CELL*>(res.rec);
```
Comment (Pospelove) [inline skymp5-server/cpp/server_guest_lib/WorldState.cpp:348]: reinterpret_cast seems to be incorrect here

it can be WRLD here. not only CELL

Comment (che6a) [reply skymp5-server/cpp/server_guest_lib/WorldState.cpp:348]: change to `espm::GetData<>` ?

Comment (Pospelove) [reply skymp5-server/cpp/server_guest_lib/WorldState.cpp:348]: espm Convert

File: .gitignore [hunk 11]
```diff
@@ -4,3 +4,4 @@ node_modules
 .vscode
 .cache
 .vs
+OFF/
```
Comment (Pospelove) [inline .gitignore:7]: why is this needed? please remove if not intended

File: skymp5-server/cpp/addon/ScampServer.cpp [hunk 12]
```diff
@@ -146,6 +149,50 @@ ScampServer::ScampServer(const Napi::CallbackInfo& info)
...
+      if (serverSettings.at("npcSettings").is_object()) {
+        std::unordered_map<std::string, WorldState::NpcSettingsEntry>
+          npcSettings;
+        for (const auto& field : serverSettings["npcSettings"].items()) {
+          WorldState::NpcSettingsEntry entry;
+          if (field.value().find("spawnInInterior") != field.value().end()) {
+            entry.spawnInInterior =
+              field.value().at("spawnInINterior").get<bool>();
```
Comment (Pospelove) [inline skymp5-server/cpp/addon/ScampServer.cpp:174]: typo spawnInINterior

File: docs/docs_server_configuration_reference.md [hunk 13]
```diff
@@ -248,3 +248,42 @@ Allows tuning settings related to in-game chat, such as message visibility radiu
...
+
+Optional npcs configuration. May not be present or can be an empty object which means all npcs are allowed to be loaded, provided `"npcEnabled"` is set to `true`.
+`"NpcSettings"` consists of fields, each of which describes from what game file it is permitted to load an npc and additional restrictions of
+how they should be spawned: in interior or exterior. By default all the npcs are allowed (`"npcSettings": {}`).
+
+```json5
+{
+  // ...
+  "npcSettings": {
+    "Skyrim.esm": {
```
Comment (Pospelove) [inline docs/docs_server_configuration_reference.md:274]: please document what happens by default.

please add a way to specify default values of spawnInInterior/spawnInExterior for non-specified esp/esm files

**Review by Pospelove — CHANGES_REQUESTED** [review]
after those fixes, testing by SupAidme will be needed

**Comment by Pospelove** [comment]
Also reformat plz

**Comment by che6a** [comment]
@Pospelove

**Review by Pospelove — APPROVED** [review]
